src/buffer_field_api.h

```diff
@@ -0,0 +1,311 @@
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+/** @file
+ * Inline API for common operations on buffer_fields
+ * @internal
+ * @{
+ */
+
+#include "qpid/dispatch/buffer_field.h"
+#include "qpid/dispatch/iterator.h"
+
+/* qd_buffer_field_memcpy
+ *
+ * Copy up to n octets from bfield to dest, advance bfield by the number of
+ * octets copied
+ *
+ * @return total of octets copied - may be < n if len(bfield) < n
+ */
+static inline size_t qd_buffer_field_memcpy(qd_buffer_field_t *bfield, uint8_t *dest, size_t n)
+{
+    const uint8_t *start = dest;
+    size_t count = MIN(n, bfield->length);
+    if (bfield->buffer) {
+        while (count) {
+            size_t avail = qd_buffer_cursor(bfield->buffer) - bfield->cursor;
+            if (count < avail) {
+                // fastpath: no need to adjust buffer pointers
+                memcpy(dest, bfield->cursor, count);
+                dest += count;
+                bfield->cursor += count;
+                bfield->length -= count;
+                return dest - start;
+            }
+
+            // count is >= what is available in the current buffer
+            memcpy(dest, bfield->cursor, avail);
+            dest += avail;
+            count -= avail;
+            bfield->length -= avail;
+            bfield->cursor += avail;
+            if (bfield->length) {
+                bfield->buffer = DEQ_NEXT(bfield->buffer);
+                if (bfield->buffer) {
+                    bfield->cursor = (const uint8_t *)qd_buffer_base(bfield->buffer);
+                } else {
+                    // DISPATCH-1394: field is truncated (length is inaccurate!)
+                    bfield->length = 0;
+                    count = 0;
+                    assert(false);  // TODO(KAG): is this fixed?
+                }
+            }
+        } while (count);
+
+        return dest - start;
+
+    } else {    // string/binary data
+
+        memcpy(dest, bfield->cursor, count);
+        bfield->cursor += count;
+        bfield->length -= count;
+        return count;
+    }
+}
+
+
+/* qd_buffer_field_advance
+ *
+ * Move the cursor of bfield forward by amount octets.
+ *
+ * @return total of octets skipped - may be < amount if len(bfield) < amount
+ */
+static inline size_t qd_buffer_field_advance(qd_buffer_field_t *bfield, size_t amount)
+{
+    size_t blen = bfield->length;
+    size_t count = MIN(amount, blen);
+    if (bfield->buffer) {
```

Comment:
Can we assume that all bfields MUST have an associated buffer. It is after all a qd_buffer_field. Don't use this data structure if you cannot associate this bfield with a buffer

Reply:
I'd like that.